core/loader/utils.ts

```diff
@@ -5,14 +5,67 @@ export function isString(attr: null | string): attr is string {
   return typeof attr === 'string' && attr.length > 0
 }
 
+type UrlSegmentType = 'relative' | 'absolute' | 'explicit'
+
+interface UrlSegment {
+  path: string
+  type: UrlSegmentType
+}
+
+function determineURLType(path: string): UrlSegmentType {
+  if (path.startsWith('/')) {
+    return 'absolute'
+  } else if (path.startsWith('https://') || path.startsWith('http://')) {
+    return 'explicit'
+  }
+  return 'relative'
+}
+
+function retrieveAllURLSegments(link: HTMLLinkElement): UrlSegment[] {
+  let parent: Element | null = link.parentElement
+  const urlSegments: UrlSegment[] = []
+
+  while (parent) {
+    const path = parent.getAttribute('xml:base')
+    parent = parent.parentElement
+    if (!path) {
+      continue
+    }
+
+    const type = determineURLType(path)
+    urlSegments.push({ path, type })
+    if (type !== 'relative') {
+      break
+    }
+  }
+  return urlSegments
+}
+
+function buildBaseURLFromSegments(
+  urlSegments: UrlSegment[],
+  baseUrl: string
+): string {
+  while (urlSegments.length > 0) {
```

Comment:
Wow, `xml:base` can be nested? 🥲

What do you think if will ignore nested `xml:base`? I think it will be extremely rare. I tried to look for `xml:base` use cases ([1](https://rssweblog.com/?guid=20050627094747), [2](https://www.odata.org/documentation/odata-version-2-0/atom-format/), [3](https://www.intertwingly.net/wiki/pie/PaceXmlBaseEverywhere), [4](https://github.com/Munter/netlify-plugin-checklinks/issues/566), [5](https://github.com/Ranchero-Software/NetNewsWire/issues/3662)) and all used absolute URL in the closes `xml:base`.

Reply:
Yea, I cried a bit as well finding that out :sweat_smile: 

Essentially there are 3 types of path segments that can go in the `xml:base`:
- absolute ( /example/ )
- explicit ( http(s)://example.com* )
- relative ( example, example/, ./example, ../example, etc...)

Absolute segments - they end the nesting and are resolving to the document base url.
Explicit segments - they end the nesting and resolve to it as the root.
Relative segments - stack until they reach one of the first 2.

If we ignore the nesting, I think that in *most* cases we will be fine.
Honestly, don't see how (and why) anyone would create the nested-routing hell. 

On the other hand, this logic handles all the cases, both "most-used" ones and all the edge ones.
If we let it say like this - and most links resolve anyway to the nearest "absolute" `xml:base` tag, we won't have any shenanigans in the traversing and stacking the route segments.

The parent-chain traversing is necessary either way, since `xml:base` may be found on any element above the link. 

Not sure which one to go with tbh. Leaning just a bit towards covering everything since it's just a 3 functions and ~50lines of code for both rss and atom loaders and resolving to the nearest "absolute" `xml:base` will have the optimal control flow either way.

But anyway, it's your call.
If you think that's an overkill, I can just redo the logic to travel to the first "absolute" path segment and end it there.
In that case, would you like to handle "explicit" segments as well or just "absolute" ones?

Reply:
I agree that it is cool to be ready for all cases. I just worry a little that the code is hard to read.

How much it will be simpler if we will cover only `absolute` case? If it will be significant simpler, let’s simplify. It will be still complicated, let’s keep it.

Reply:
(By the way, thanks for deep investigation 👍)

Reply:
> I just worry a little that the code is hard to read.

Yea, my biggest concern was that as well. It's too hacky.

Let me sleep on it and see how far can I simplify for both approaches.

Reply:
@ai , updated the PR. Should be much cleaner now.

Reply:
Yes, looks great